lib/gc/Transforms/DeepTileContractionNamedOp.cpp

```diff
@@ -0,0 +1,1005 @@
+  if (!permutation.empty()) {
+    SmallVector<int64_t> transposeShape;
+    for (int64_t idx : permutation)
+      transposeShape.push_back(outShape[idx]);
+    Operation *initOp = rewriter.create<tensor::EmptyOp>(loc, transposeShape,
+                                                         tensorElementType);
+    Operation *transposeOp = rewriter.create<linalg::TransposeOp>(
+        loc, result, initOp->getResult(0), permutation);
+    result = transposeOp->getResult(0);
+  }
+  return result;
+}
+
+// Check if the loop is dummy loop(has only one iteration)
+bool isDummyLoop(LoopLikeOpInterface loop) {
+  std::optional<int64_t> tripCount = mlir::constantTripCount(
+      *loop.getSingleLowerBound(), *loop.getSingleUpperBound(),
+      *loop.getSingleStep());
+  if (tripCount)
+    return *tripCount == 1;
+  return false;
+}
+
+// Build the linalg region for a linalg op
+static void buildLinalgRegion(Operation *op, bool createTemporaryOp = false) {
+  SmallVector<Type> argTypes;
+  SmallVector<Location> argLocs;
+  for (const Value &opOperand : op->getOperands()) {
+    argTypes.push_back(getElementTypeOrSelf(opOperand.getType()));
+    argLocs.push_back(opOperand.getLoc());
+  }
+  size_t initSize = op->getResults().size();
+  ImplicitLocOpBuilder b(op->getLoc(), op->getContext());
+  Region &region = op->getRegion(0);
+  Block *body = b.createBlock(&region, /*insertPt=*/{}, argTypes, argLocs);
+  b.setInsertionPointToStart(body);
+  if (createTemporaryOp) {
+    unsigned argNum = body->getNumArguments();
+    SmallVector<Value> vals;
+    for (size_t i = initSize; i > 0; i--)
+      vals.push_back(body->getArgument(argNum - i));
+    OpBuilder::InsertionGuard g(b);
+    b.setInsertionPointToEnd(body);
+    Location loc = b.getUnknownLoc();
+    b.create<linalg::YieldOp>(loc, ValueRange(vals));
+  } else {
+    linalg::LinalgDialect *dialect =
+        static_cast<linalg::LinalgDialect *>(op->getDialect());
+    linalg::LinalgDialect::RegionBuilderFunType fun =
+        dialect->getRegionBuilder("linalg.matmul");
+    fun(b, *body, op->getAttrs());
+  }
+}
+
+// Check if the linalgOp need to be legalized to f32 accumulation type
+static bool needToLegalizeDtype(linalg::LinalgOp linalgOp) {
+  mlir::Type dataType =
+      dyn_cast<mlir::ShapedType>(linalgOp.getDpsInputs()[0].getType())
+          .getElementType();
+  mlir::Type resultType =
+      dyn_cast<mlir::ShapedType>(linalgOp.getDpsInits()[0].getType())
+          .getElementType();
+  return (dataType.isBF16() || dataType.isF16()) && dataType == resultType;
+}
+
+struct DtypeLegalizeResult {
+  Operation *linalgOp = nullptr;
+  Operation *castOp = nullptr;
+};
+
+// Split a low precision matmul(bf16xbf16->bf16) to a combination
+// matmul(bf16xbf16->f32) + cast(f32->bf16)
+// if needFurtherFuse=true, a middle temporary linalgOp(bf16xbf16->(f32,bf16))
+// will be created
+static FailureOr<DtypeLegalizeResult>
+matmulDtypeLegalize(RewriterBase &rewriter, Operation *op,
```

Comment:
It is about the accumulation type. Matmul(bf16xbf16->bf16) will do accumulation on bf16 while matmul(bf16xbf16->f32)+cast(f32->bf16) will do accumulation on f32 and cast back to bf16.

Reply:
I was referring to having different in & out datatype for matmul ops, so that the fusion will handle the case like  `linalg.matmul(bf16:bf16:f32) + tensor.cast()` instead of an internal pass to handle this?

Reply:
The current fusion cannot fuse the cast into the innermost loop if the K/reduction axis is not full. So the `deepTileMatmul` will temporarily fuse the cast into the innermost loop in a template way.

```
deep tile matmul:
for(M : MBlocks) {
    for(K: KBlocks) {
        for(N: NBlocks) {
            C = brgemm(M, K, N);
            cast_to_bf16(C);
        }
    }
}
current fusion: 
for(M : MBlocks) {
    for(K: KBlocks) {
        for(N: NBlocks) {
            C = brgemm(M, K, N);
        }
    }
    cast_to_bf16(bigger_C);
}
```

Reply:
I see, so this is one of the inefficiency of automatically fusion approach. Thanks for the explaination.